src/BlockHorizons/PerWorldPlayer/world/WorldInstance.php

```diff
@@ -61,14 +71,27 @@ public function onPlayerEnter(Player $player, ?WorldInstance $from_world = null)
 	}
 
 	public function onPlayerExit(Player $player, ?WorldInstance $to_world = null, bool $quit = false) : void{
-		if($to_world === null || !self::haveSameBundles($this, $to_world)){
-			$this->save($player, PlayerWorldData::fromPlayer($player), false, $quit);
+		if($to_world === null || !self::haveSameBundles($this, $to_world)){ //TODO: currently plugins cannot bypass this
+			$this->save($player, PlayerWorldData::fromPlayer($player), false, $quit ? PerWorldPlayerDataSaveEvent::CAUSE_PLAYER_QUIT : PerWorldPlayerDataSaveEvent::CAUSE_WORLD_CHANGE);
 		}
 	}
 
-	public function save(Player $player, PlayerWorldData $data, bool $force = false, bool $quit = false) : void{
-		if($force || !$player->hasPermission("per-world-player.bypass")){
-			$this->database->save($this, $player, $data, $quit);
+	public function save(Player $player, PlayerWorldData $data, bool $force = false, int $cause = PerWorldPlayerDataSaveEvent::CAUSE_CUSTOM) : void{
```

Comment:
Preferably the order of parameters here as `($player, $data, $cause, $force)` would be nice as more likely than not, the $force would be lesser used as it overwrites default behaviour.